src/network/core/os_abstraction.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @file os_abstraction.cpp OS specific implementations of functions of
+ *                          the OS abstraction layer for network stuff.
+ *
+ * The general idea is to have simple abstracting functions for things that
+ * require different implementations for different environments.
+ * In here the functions, and their documentation, are defined only once
+ * and the implementation contains the #ifdefs to change the implementation.
+ * Since Windows is usually different that is usually the first case, after
+ * that the behaviour is usually Unix/BSD-like with occasional variation.
+ */
+
+#include "stdafx.h"
+#include "os_abstraction.h"
+#include "../../string_func.h"
+
+#include "../../safeguards.h"
+
+/**
+ * Construct the network error with the given error code.
+ * @param error The error code.
+ */
+NetworkError::NetworkError(int error) : error(error)
+{
+}
+
+/**
+ * Construct the network error with the last available error.
+ */
+NetworkError::NetworkError()
+{
+#if defined(_WIN32)
+	this->error = WSAGetLastError();
+#elif defined(__OS2__)
+	this->error = sock_errno();
+#else
+	this->error = errno;
+#endif
+}
+
+/**
+ * Check whether this error describes that the operation would block.
+ * @return True iff the operation would block.
+ */
+bool NetworkError::WouldBlock() const
+{
+#if defined(_WIN32)
+	return this->error == WSAEWOULDBLOCK;
+#elif EWOULDBLOCK != EAGAIN
+	/* Usually EWOULDBLOCK and EAGAIN are the same, but sometimes they are not
+	 * and the POSIX.1 specification states that either should be checked. */
+	return this->error == EWOULDBLOCK || this->error == EAGAIN;
+#else
+	return this->error == EWOULDBLOCK;
+#endif
```

Comment:
Can we not assume that the compiler will optimise `this->error == FOO || this->error == FOO` ? Seems a bit unnecessary to have both implementations (but keep the comment)

Reply:
I had copied the idea of doing it separately from TB's proposal without much thought, except that it made some sense as I would expect (pedantic) compilers to complain about comparing the same thing twice. Though maybe they're not as pedantic as I expected them to be.